Add an optimization that removes the inner multiply in dsiAccess

ronawho opened the pull request:
Add a compiler optimization (--optimize-array-indexing) that removes the inner
multiply for array indexing when possible. The basic idea of the optimization
is that if the blk field is never modified, we know that blk(rank) will always
be 1, so we can optimize away the last blk multiply in dsiAccess().

In order to determine if blk is modified, we explicitly mark all functions that
might modify it with a pragma. Then, if none of those functions are called we
know blk cannot be modified, so it's safe to remove the multiply. The functions
that modify blk are adjustBlkOffStrForNewDomain() (called for strided reindexing
and strided slicing) and rankChange().

This is kind of a lame optimization because if any array requires blk, all
arrays will suffer the performance penalty. i.e. the optimization just checks
if any function modifies block, it doesn't do any contextual or type based
analysis. Longer term we expect array views will provide a way to remove the
inner multiply in a cleaner and more principled manner, so this optimization is
really just a stopgap until that comes along.

Until then, this gets us the performance benefit for most cases, particularly
for benchmarks like the PRKs, array microbenchmarks, lcals, and some of the
shootout benchmarks. It also allows us to remove the hackier assertNoSlicing
param, because we now optimize almost all the cases where assertNoSlicing
wouldn't break things anyways.

The generated code improvement from this optimization is also pretty sweet:

```chapel
// stream like code
for i in 1..10 do A[i] = B[i] + C[i];
```

used to produce:

```c
for (i = INT64(1); i <= INT64(10); i += INT64(1)) {
  sum = INT64(0);                         // dead store
  sum += ((int64_t)((i * coerce_tmp)));   // extra mult
  call_tmp7 = (coerce_tmp2 + sum);
  sum2 = INT64(0);
  sum2 += ((int64_t)((i * coerce_tmp3)));
  call_tmp9 = (coerce_tmp4 + sum2);
  ret = *(call_tmp9);
  sum3 = INT64(0);
  sum3 += ((int64_t)((i * coerce_tmp5)));
  call_tmp11 = (coerce_tmp6 + sum3);
  *(call_tmp7) = ((int64_t)((ret + *(call_tmp11))));
}
```

and now generates: 

```c
for (i = INT64(1); i <= INT64(10); i += INT64(1)) {
  call_tmp7 = (coerce_tmp + i);
  call_tmp9 = (coerce_tmp2 + i);
  call_tmp11 = (coerce_tmp3 + i);
  *(call_tmp7) = ((int64_t)((*(call_tmp9) + *(call_tmp11))));
}
```

ronawho commented:
@bradcray there's still some work I have to do (noted in the TODOs above) but this is what the actual optimization looks like. 

@benharsh  I'm curious to see how much you don't like this

benharsh commented:
You and Brad made good points that this results in some good performance improvements and removes the need for users to throw some odd flag, so I'm no longer opposed :+1:

benharsh commented:
This looks good to me. What testing has been done?

ronawho commented:
std testing as well as gasnet, plus we've been testing in the perf playground for a while

bradcray commented on compiler/main/driver.cpp:
```diff
@@ -646,6 +650,8 @@ static ArgumentDescription arg_desc[] = {
  {"inline-iterators", ' ', NULL, "Enable [disable] iterator inlining", "n", &fNoInlineIterators, "CHPL_DISABLE_INLINE_ITERATORS", NULL},
  {"live-analysis", ' ', NULL, "Enable [disable] live variable analysis", "n", &fNoLiveAnalysis, "CHPL_DISABLE_LIVE_ANALYSIS", NULL},
  {"loop-invariant-code-motion", ' ', NULL, "Enable [disable] loop invariant code motion", "n", &fNoloopInvariantCodeMotion, NULL, NULL},
+
```
Minor, but is this blank line intentional?  Does it have any effect?

ronawho replied:
Nope, not intentional. Thanks

bradcray commented on modules/internal/DefaultRectangular.chpl:
```diff
@@ -842,24 +841,31 @@ module DefaultRectangular {
+          if __primitive("optimize_array_blk_mult") {
+            return ind(1);
+          } else {
+            return ind(1) * blk(1);
```
Does breaking out the rank==1 case like this help substantially?

ronawho replied:
I didn't see any performance impact from it, but it has a significant impact on the generated code readability

```chapel
for i in 1..10 do A[i] = B[i] + C[i];
```

```c
  // w/o rank == 1
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = INT64(0);
    sum += i;
    call_tmp7 = (coerce_tmp + sum);
    sum2 = INT64(0);
    sum2 += i;
    call_tmp9 = (coerce_tmp2 + sum2);
    ret = *(call_tmp9);
    sum3 = INT64(0);
    sum3 += i;
    call_tmp11 = (coerce_tmp3 + sum3);
    *(call_tmp7) = ((int64_t)((ret + *(call_tmp11))));
  }
```
vs 
```c
  //w/ rank == 1
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    call_tmp7 = (coerce_tmp + i);
    call_tmp9 = (coerce_tmp2 + i);
    call_tmp11 = (coerce_tmp3 + i);
    *(call_tmp7) = ((int64_t)((*(call_tmp9) + *(call_tmp11))));
  }
```

The backend should have no problem optimizing what we would generate without the rank==1 special case, but I think the generated code improvement is worthwhile.

ronawho replied:
I'm open to dropping the rank ==1 specialization if you're strongly against it, or I'm also open to doing it in another PR if you think that would be better

bradcray replied:
I'm definitely OK with leaving it in, particularly with the rationale noted here (though noting it in the comments would be even better.  

It does make me wonder whether there would be similar improvements we could do to the multi-dim case...  I assume all of this was with denormalization off?

Anyway, no need to do anything beyond a comment in the code for my sake.

ronawho replied:
good point, I'll add a comment. And actually, that was with dernom turned on.


We can do the same sort of thing for multidim, but the generated code for that is already much uglier and it's hard to notice the improvement


```chapel
  for i in 1..10 do A[i, i] = i;
```

```c
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = INT64(0);
    sum += ((int64_t)((i * coerce_tmp)));
    sum += i;
    call_tmp8 = (coerce_tmp2 + sum);
    *(call_tmp8) = i;
  }
```

vs

```c
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = ((int64_t)((i * coerce_tmp)));
    sum += i;
    call_tmp8 = (coerce_tmp2 + sum);
    *(call_tmp8) = i;
  }
```

bradcray commented:
I looked only at a high level, but it was essentially what I expected and I'm happy with it (modulo more careful review from someone else, which may have been @benharsh).

One thing I'm wondering is whether it would be possible to put in a test (a future?) that demonstrates how reindexing one array infects other unrelated arrays as motivation for more precise optimization in the future (e.g., through array views).

The experimental part of me would be curious to deprecate assertNoSlicing first, get a night of performance runs, and then put the rest of this in the following night.  But since time's short before feature freeze, probably we should just go for it.

ronawho commented:
>One thing I'm wondering is whether it would be possible to put in a test (a future? or a perf test?) that demonstrates how reindexing one array infects other unrelated arrays as motivation for more precise optimization in the future (e.g., through array views).

I'm not really sure what such a test would even do short of grepping the generated code, which I really don't want todo. Would opening a JIRA issue be good enough?

>The experimental part of me would be curious to deprecate assertNoSlicing first, get a night of performance runs, and then put the rest of this in the following night. But since time's short before feature freeze, probably we should just go for it.

Yeah, I was wondering the same thing, but I think in the long term that would just make the graphs noisier, and as you said we're running short on time

bradcray commented:
I edited my comment belatedly -- what about a perf test that timed array accesses of two arrays, one needing the inner mult, the other not in a tight loop?  This could go in separately / later.

ronawho commented:
>I edited my comment belatedly -- what about a perf test that timed array accesses of two arrays, one needing the inner mult, the other not in a tight loop? This could go in separately / later.

Ah, as a perf test makes sense. Yeah, I'll go head and add that (probably post-release though, since I'm hoping to get back to fast-followers today yet)

ronawho commented:
Added rank == 1 comment, removed extra newline in driver.cpp, updated PR message, and cleaned up the history. Doing  a final test run now before merging. Let me know if you have any other comments
